### PR TITLE
[DI] Exclude private services from alternatives

### PR DESCRIPTION
**Pull request by ro0NL**
| Q | A |
| --- | --- |
| Branch? | "master" |
| Bug fix? | yes |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | yes |
| Tests pass? | yes |
| Fixed tickets | comma-separated list of tickets fixed by the PR, if any |
| License | MIT |
| Doc PR | reference to the documentation PR, if any |

After #19685 & #19690

### CODE REVIEW

**Comment by ogizanagi** [comment]
This should go into `2.7`, not `master`.

However, you won't be able to exclude privates as `Container::privates` is only available since 3.2.

File: src/Symfony/Component/DependencyInjection/Tests/ContainerTest.php [hunk 1]
```diff
@@ -198,6 +197,23 @@ public function testGetThrowServiceNotFoundException()
...
+        $sc = new ProjectServiceContainer();
+        $sc->set('internal_lookalike', $foo = new \stdClass());
+        $privates = new \ReflectionProperty(ProjectServiceContainer::class, 'privates');
+        $privates->setAccessible(true);
+        $this->assertArrayHasKey('internal', $privates->getValue($sc));
+
+        try {
+            $sc->get('int');
+            $this->fail('->get() throws an Symfony\Component\DependencyInjection\Exception\ServiceNotFoundException if the key does not exist');
+        } catch (\Exception $e) {
```
Comment (ogizanagi) [inline src/Symfony/Component/DependencyInjection/Tests/ContainerTest.php]: Why not using `@expectedException` & `@ExpectedExceptionMessage` annotations ?

Comment (ro0NL) [inline src/Symfony/Component/DependencyInjection/Tests/ContainerTest.php]: Copy paste :)

**Comment by ro0NL** [comment]
@ogizanagi could be done :) i dont mind.

**Comment by ogizanagi** [comment]
@ro0nl: I'll suggest you to create another PR for 2.7 and up, and keep this one to not loose track of the private services handling in 3.2.

File: src/Symfony/Component/DependencyInjection/Container.php [hunk 2]
```diff
@@ -270,6 +270,9 @@ public function get($id, $invalidBehavior = self::EXCEPTION_ON_INVALID_REFERENCE
 
                     $alternatives = array();
                     foreach ($this->services as $key => $associatedService) {
+                        if (isset($this->privates[$key])) {
```
Comment (ogizanagi) [inline src/Symfony/Component/DependencyInjection/Container.php]: What about moving this check into `getServiceIds()` and add an optional `$private = true` argument (which will be deprecated later, in order to the `getServiceIds()` to only return public services ids ?

File: src/Symfony/Component/DependencyInjection/Tests/ContainerTest.php (hunk 1 above)
Comment (ogizanagi) [inline src/Symfony/Component/DependencyInjection/Tests/ContainerTest.php]: IMHO, unless there is a good reason, new tests should use the annotations.
When older tests where introduced, I guess those annotations (as well as `TestCase::setExpectedException()`) were not available with the PHPUnit version in use :)

Comment (fabpot) [inline src/Symfony/Component/DependencyInjection/Tests/ContainerTest.php]: Indeed, new tests should use the annotations.

File: src/Symfony/Component/DependencyInjection/Container.php (hunk 2 above)
Comment (ro0NL) [inline src/Symfony/Component/DependencyInjection/Container.php]: Makes sense for having `getServiceIds` only expose public id's.

Not sure why we want `$private = true`. Opposed to refactoring stuff relying on the fact it returns private ids.

Comment (ogizanagi) [inline src/Symfony/Component/DependencyInjection/Container.php]: > Not sure why we want $private = true. Opposed to refactoring stuff relying on the fact it returns private ids.

That's only in order to ensure BC. Having this argument set to true should trigger a deprecation. Internally, we'll only use `false`. (if any internal call needs privates services, we'll have to add another `$triggerDeprecation = true` argument and set it to `false`, but I doubt there is.)

Comment (ro0NL) [inline src/Symfony/Component/DependencyInjection/Container.php]: > Yes. Why did you changed ?

Waiting for upstream to change :stuck_out_tongue:, but ill add it. Makes sense.

Got your point about `getServiceIds`. It's public API.. which should keep returning privates till 4.0. Ill have a look at it.

File: src/Symfony/Component/DependencyInjection/Container.php [hunk 3]
```diff
@@ -339,11 +336,22 @@ public function reset()
...
      * @return array An array of all defined service ids
      */
-    public function getServiceIds()
+    public function getServiceIds(/*$includePrivates = true*/)
     {
+        $includePrivates = true;
+        if (func_num_args() === 1) {
+            $includePrivates = !!func_get_arg(0);
+            if ($includePrivates) {
+                @trigger_error(sprintf('Including private services in the list of service id\'s is deprecated since Symfony 3.2 and won\'t be supported anymore in Symfony 4.0.', $id), E_USER_DEPRECATED);
```
Comment (ro0NL) [inline src/Symfony/Component/DependencyInjection/Container.php]: @ogizanagi something like this?

Comment (ogizanagi) [inline src/Symfony/Component/DependencyInjection/Container.php]: Sorry, I don't understand your concerns 😕 ?

You can't just deprecate privates in `getServiceIds` if you have no idea the privates were asked. And simply not returning them will be a BC break. Thus, the new argument to explicit the intention.

Comment (ro0NL) [inline src/Symfony/Component/DependencyInjection/Container.php]: Got it. `master != 4.x` this keeps confusing me :sweat_smile:

File: src/Symfony/Component/DependencyInjection/Container.php [hunk 4]
```diff
@@ -336,11 +336,22 @@ public function reset()
      *
      * @return array An array of all defined service ids
      */
-    public function getServiceIds()
+    public function getServiceIds(/*$includePrivates = true*/)
     {
+        $includePrivates = true;
+        if (func_num_args() === 1) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Container.php]: Nobody calls this with an argument, which means nobody will see the deprecation notice.
Yet, I think we should only add a note to remind about not returning private services in 4.0. No need for any deprecations: if people use the private services returned today from this method, they will get a deprecation because fetching them is deprecated.

File: src/Symfony/Component/DependencyInjection/Container.php [hunk 5]
```diff
@@ -269,10 +269,10 @@ public function get($id, $invalidBehavior = self::EXCEPTION_ON_INVALID_REFERENCE
                     }
 
                     $alternatives = array();
-                    foreach ($this->services as $key => $associatedService) {
-                        $lev = levenshtein($id, $key);
-                        if ($lev <= strlen($id) / 3 || false !== strpos($key, $id)) {
-                            $alternatives[] = $key;
+                    foreach ($this->getServiceIds(false) as $knownId) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Container.php]: this change is unrelated and should be discussed in #19685 (thus no need for the patch here)

Comment (ro0NL) [inline src/Symfony/Component/DependencyInjection/Container.php]: Agree. Im reverting it so we check for privates here again.

Deprecating privates from `getServiceIds` is a new PR i guess.. it needs to be taken care of.

Comment (ro0NL) [inline src/Symfony/Component/DependencyInjection/Container.php]: @nicolas-grekas if #19685 and a new PR, fixing excluding privates from `getServiceIds`, are done.... this becomes a non-issue. Shall we close it?

Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Container.php]: I think this PR should be closed yes.

**Comment by ro0NL** [comment]
Superseded by #19707 + #19685
